### PR TITLE
Support sorting invocations on the history page.

### PR DESCRIPTION
**Pull request by iain-macdonald**
This PR adds a new chip to the history page for sorting invocations. I added a handful of sort options including:
- Invocation start time
- Invocation end time
- Invocation duration
- Action Cache hit ratio
- Content Addressable Store hit ratio
- Cache download
- Cache upload
- Cache transfer

It supports sorting by any of these in ascending or descending order. It should be pretty easy to add new sort parameters in the future too. Here are some screenshots:

Screenshot 2022-11-30 at 8 33 07 AM
Screenshot 2022-11-30 at 8 33 29 AM
Screenshot 2022-11-30 at 8 33 40 AM

---

**Version bump**: Minor
**Related issues**: Fixes #1223

### CODE REVIEW

**Comment by siggisim** [comment]
This is awesome! I'll let @bduffany do the review - but what do you think about putting the dropdown in between the filter and date dropdowns? That way filter and sort are next to one another (which feel the most closely related).

Also wondering if the default button shouldn't have any text (just the sort icon) until the default sort order is changed (maybe this is already the case!) so it doesn't take up as much space.

File: enterprise/server/invocation_search_service/invocation_search_service.go
Comment (luluz66) [inline enterprise/server/invocation_search_service/invocation_search_service.go:188]: I'm a bit concerned about the query performance, since these fields are not indexed. Have you profiled these queries or tested against our biggest customers?

File: enterprise/app/filter/filter.tsx
Comment (bduffany) [inline enterprise/app/filter/filter.tsx:208]: nit: replace `Toggle` with `Change` in these 2 function names since it's changing the value of a radio selection rather than toggling a single checkbox element

File: app/router/router.tsx
Comment (bduffany) [inline app/router/router.tsx:25]: (to match the consts above)

```suggestion
export const SORT_BY_PARAM_NAME = "sort-by";
export const SORT_ORDER_PARAM_NAME = "sort-order";
```

File: enterprise/app/filter/filter.tsx
Comment (bduffany) [inline enterprise/app/filter/filter.tsx:461]: nit: in the case where the button text is empty, it looks a bit too wide. It would be nice to have it look nice and square like the filter icon. When the text is empty, you can use the class `icon-button` instead of `icon-text-button` and remove the empty `<span>` element.

**Review by bduffany — CHANGES_REQUESTED** [review]

File: enterprise/app/filter/filter.tsx
Comment (siggisim) [reply enterprise/app/filter/filter.tsx:466]: Cool with anything here! "Start time" and "End time" sounds reasonable

File: enterprise/app/filter/filter.tsx
Comment (iain-macdonald) [reply enterprise/app/filter/filter.tsx]: Good call, removed.

File: enterprise/app/filter/filter.tsx
Comment (iain-macdonald) [reply enterprise/app/filter/filter.tsx]: Whoops, thanks for pointing that out.

File: enterprise/app/history/history.tsx
Comment (iain-macdonald) [reply enterprise/app/history/history.tsx]: Done.

File: enterprise/app/filter/filter.tsx
Comment (iain-macdonald) [reply enterprise/app/filter/filter.tsx]: Done.

File: enterprise/app/filter/filter.tsx
Comment (iain-macdonald) [reply enterprise/app/filter/filter.tsx]: Yeah I prefer those as well, changed.

File: enterprise/app/filter/filter.tsx
Comment (iain-macdonald) [reply enterprise/app/filter/filter.tsx]: Done.

File: enterprise/app/filter/filter.tsx
Comment (iain-macdonald) [reply enterprise/app/filter/filter.tsx]: Good call, fixed.

File: enterprise/server/invocation_search_service/invocation_search_service.go
Comment (iain-macdonald) [reply enterprise/server/invocation_search_service/invocation_search_service.go:188]: Thanks for pointing that out. In the common case (no sorting), the query to the database will be the same, so that shouldn't be affected by this PR. I suspect you're right that some of the other queries will be slower, but I don't think we should block this PR because this is a new feature and we can look into speeding it up if it proves to be a problem. Does that seem reasonable?

**Review by iain-macdonald — COMMENTED** [review]
Thanks, Brandon & Lulu! I think I addressed all your comments, PTAL.

**Comment by bduffany** [comment]
It looks like there are a few typecheck failures -- we removed type checking from the build and made them tests instead, since TS type checking is super slow. To run the type check tests locally you can do

```
bb test //enterprise/app/... //app/...
```

File: enterprise/app/filter/filter.tsx
Comment (bduffany) [inline enterprise/app/filter/filter.tsx:315]: TS will type these as string, so to get the type checking benefits you can change this to

```suggestion
    const sortByValue: SortBy | "" = (this.props.search.get(SORT_BY_PARAM_NAME) as SortBy) || "";
    const sortOrderValue: SortOrder = (this.props.search.get(SORT_ORDER_PARAM_NAME) as SortOrder) || "desc";
```

File: enterprise/app/history/history.tsx
Comment (bduffany) [inline enterprise/app/history/history.tsx:123]:
```suggestion
        ascending: filterParams.sortOrder === "asc",
```

File: enterprise/app/filter/filter.tsx
Comment (iain-macdonald) [reply enterprise/app/filter/filter.tsx]: Whoops, done. I went ahead and added "" as a valid SortBy value too so sortByValue can be just a SortBy here.

File: enterprise/server/invocation_search_service/invocation_search_service.go
Comment (iain-macdonald) [reply enterprise/server/invocation_search_service/invocation_search_service.go:199]: Whoops, forgot to comment on this...

I thought about this, and added the special divide-by-zero logic to deal with those items being ranked weird. I figured that for now this is probably good enough and if we hear feedback we can update this calculation to use a weight average or something more sophisticated.

**Review by iain-macdonald — COMMENTED** [review]
Thanks, Brandon, PTAL!

**Review by bduffany — APPROVED** [review]
Awesome!

File: enterprise/app/filter/filter.tsx
Comment (bduffany) [inline enterprise/app/filter/filter.tsx:475]: (woops, forgot to add this in my previous review)

nit: the icon doesn't appear centered because of the empty `<span>` in the case where none of the labels is rendered.

Comment (iain-macdonald) [reply enterprise/app/filter/filter.tsx]: Oh shoot, you left this comment in an earlier review and I accidentally ignored it. Fixed and updated screenshot in PR with the fix.

**Pull request merged**
